### PR TITLE
Fix deprecation warnings

### PR DESCRIPTION
**Pull request by lsh-0**
- [x] api.py shouldn't have been working with `listfiles` broken. tests!
- [x] review

### CODE REVIEW

**Comment by lsh-0** [comment]
regarding removing `connexion`:

> it had a deprecation warning that was only fixed in the next (unreleased) major version.

turns out it was also responsible for these deprecation warnings as well:

```
src/tests/test_api.py: 53 warnings
  /home/luke/dev/python/bot-lax-adaptor/venv/lib/python3.8/site-packages/flask/json/provider.py:188: DeprecationWarning: Setting 'json_encoder' on the app or a blueprint is deprecated and will be removed in Flask 2.3. Customize 'app.json' instead.
    warnings.warn(

src/tests/test_api.py: 53 warnings
  /home/luke/dev/python/bot-lax-adaptor/venv/lib/python3.8/site-packages/flask/json/provider.py:230: DeprecationWarning: 'JSONEncoder' is deprecated and will be removed in Flask 2.3. Use 'Flask.json' to provide an alternate JSON implementation instead.
    return json.dumps(obj, **kwargs)
```

removing `connexion` mopped up the last of the warnings and this PR is ready to be reviewed.

File: src/utils.py [hunk 1]
```diff
@@ -296,3 +310,28 @@ def msid_from_elife_doi(doi):
...
+    except ValueError as err:
+        LOG.error("struct is not serializable: %s", str(err))
+        raise
+
+    try:
+        validator(struct, schema, format_checker=jsonschema.FormatChecker())
+        return struct
+
+    except ValueError as err:
+        # your json is broken
```
Comment (lsh-0) [inline src/utils.py:331]: rm 'your'

File: src/utils.py [hunk 2]
```diff
@@ -296,3 +310,28 @@ def msid_from_elife_doi(doi):
...
+        raise ValidationError("validation error: '%s' for: %s" % (str(err), struct))
+
+    except ValidationError as err:
+        # your json is incorrect
```
Comment (lsh-0) [inline src/utils.py:335]: rm 'your'

File: src/utils.py [hunk 3]
```diff
@@ -296,3 +310,28 @@ def msid_from_elife_doi(doi):
         return None
     regex = r"10.7554/elife\.(?P<msid>\d+)"
     return first(re.findall(regex, doi, re.IGNORECASE))
+
+def validate(struct, schema):
+    # if given a string, assume it's json and try to load it
```
Comment (lsh-0) [inline src/utils.py:315]: proper docstr

File: src/utils.py [hunk 4]
```diff
@@ -272,10 +255,41 @@ def target(*args, **kwargs):
...
+    if not isinstance(dt, datetime):
+        raise AssertionError("given value is not a datetime.datetime object: %r" % dt)
+
+    dt.replace(microsecond=0) # not useful, never been useful, will never be useful.
+
+    # no tz. assume UTC and make it explicit.
+    if not dt.tzinfo:
+        return pytz.utc.localize(dt)
+
+    # has tz, but it's not utc. ensure tz is UTC.
```
Comment (lsh-0) [inline src/utils.py:279]: uppercase UTC

File: src/utils.py [hunk 5]
```diff
@@ -272,10 +255,41 @@ def target(*args, **kwargs):
     return resp
 
 def todt(val):
-    return dateutils.todt(val)
+    "turn almost any formatted datetime string into a UTC datetime object"
+    if val is None:
+        return None
+    dt = val
```
Comment (lsh-0) [inline src/utils.py:261]: use `dt` or `val`

File: src/utils.py [hunk 6]
```diff
@@ -272,10 +255,41 @@ def target(*args, **kwargs):
     return resp
 
 def todt(val):
-    return dateutils.todt(val)
+    "turn almost any formatted datetime string into a UTC datetime object"
```
Comment (lsh-0) [inline src/utils.py:258]: update docstr

File: src/utils.py [hunk 7]
```diff
@@ -256,7 +239,7 @@ def target(*args, **kwargs):
         response = s.send(prepared_request)
         if response.status_code >= 500:
             raise RemoteResponseTemporaryError("Status code was %s" % response.status_code)
-        #dumpobj((request, response))
+        # dumpobj((request, response))
```
Comment (lsh-0) [inline src/utils.py:242]: rm

File: src/utils.py [hunk 8]
```diff
@@ -119,33 +121,14 @@ def first(x):
...
-        LOG.error("struct failed to validate against schema: %s" % str(err))
-        raise
-
 def json_dumps(obj, **kwargs):
-    return dateutils.json_dumps(obj, **kwargs)
+    "drop-in for json.dumps that handles datetime objects."
+    def datetime_handler(obj):
+        if hasattr(obj, 'isoformat'):
+            return ymdhms(obj)
+        else:
```
Comment (lsh-0) [inline src/utils.py:129]: rm `else`

File: src/utils.py [hunk 9]
```diff
@@ -119,33 +121,14 @@ def first(x):
...
-        raise ValidationError("validation error: '%s' for: %s" % (str(err), struct))
-
-    except ValidationError as err:
-        # your json is incorrect
-        LOG.error("struct failed to validate against schema: %s" % str(err))
-        raise
-
 def json_dumps(obj, **kwargs):
-    return dateutils.json_dumps(obj, **kwargs)
+    "drop-in for json.dumps that handles datetime objects."
```
Comment (lsh-0) [inline src/utils.py:125]: ticks

File: src/tests/test_utils.py [hunk 10]
```diff
@@ -149,7 +149,7 @@ def test_sortdict(self):
         for given, expected in cases:
             self.assertEqual(utils.json_dumps(expected), utils.json_dumps(utils.sortdict(given)))
             # print(utils.sortdict(given))
-            #utils.ensure(expected == utils.sortdict(given), "%s != %s" % (expected, utils.sortdict(given)))
+            # utils.ensure(expected == utils.sortdict(given), "%s != %s" % (expected, utils.sortdict(given)))
```
Comment (lsh-0) [inline src/tests/test_utils.py:152]: rm these comments

File: src/api.py [hunk 11]
```diff
@@ -177,7 +150,7 @@ def post_xml():
 
     # send to lax
     try:
-        #msid, version = utils.version_from_path(filename)
+        # msid, version = utils.version_from_path(filename)
```
Comment (lsh-0) [inline src/api.py:153]: rm

File: src/adaptor.py [hunk 12]
```diff
@@ -90,7 +90,7 @@ def find_lax():
     return script
 
 def call_lax(action, msid, version, token, article_json=None, force=False, dry_run=False):
-    #raise EnvironmentError("whoooooa. no.")
+    # raise EnvironmentError("whoooooa. no.")
```
Comment (lsh-0) [inline src/adaptor.py:93]: rm

File: src/utils.py [hunk 13]
```diff
@@ -272,10 +255,41 @@ def target(*args, **kwargs):
...
+
+    return dt
 
 def ymdhms(dt):
-    return dateutils.ymdhms(dt)
+    "returns an rfc3339 (2023-03-01T04:48:16Z) representation of a datetime object."
+    if not dt:
+        return
+    if not isinstance(dt, datetime):
+        raise AssertionError("given datetime value is not a datetime.datetime object: %r" % dt)
```
Comment (lsh-0) [inline src/utils.py:290]: TypeError

**Comment by lsh-0** [comment]
yeah, `listfiles` had no coverage at all:

Screenshot at 2023-03-06 16-13-10

it looks like it was only breaking the GET endpoints for `/xml` and `/article-json/*`, so the much more important POST to `/xml` has been working.

**Comment by lsh-0** [comment]
@gnott , I'm probably going to merge this tomorrow. I've added you as a reviewer but don't feel obliged, it's more of an FYI.

I've gone through the dependency list and removed some unused or mostly unnecessary dependencies to fix deprecation warnings.

I've also done some housecleaning/linting so the PR is a little noisy.

**Review by gnott — APPROVED** [review]
I had a read-through, I think I can see the main results of refactoring, using `Flask`, updating the dependencies.

**Comment by lsh-0** [comment]
the main changes were:
* remove connexion
* remove flex
* remove yaml
* replace aws requests signing
* merged dateutils with utils

**Pull request merged**